### DISCUSSION TITLE
Do we solely rely on memory management to call UploadFile.close?

### DISCUSSION THREAD

**Discussion by originell** [fastapi/fastapi#6380] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# From the docs https://fastapi.tiangolo.com/tutorial/request-files/#import-file
from fastapi import FastAPI, File, UploadFile

app = FastAPI()


@app.post("/uploadfile/")
async def create_upload_file(file: UploadFile):
    return {"filename": file.filename}
```

### Description

I realise this might be a fairly in-depth question and I'm thankful for anyone reading this in its entirety.

My question is about handling of file descriptors: why do the docs not use `await file.close()` explicitly? 

When handling files in Python, one is usually encouraged to either use the old `try/finally` pattern or a context manager. e.g.:

```python
# pre-context manager style
try:
    img = open("me.png", "rb")
    # do something with it that might error out
    print(img)
finally:
    img.close()

# or the new style
with open("me.png", "rb") as img:
    # process. img.close() called by context manager when leaving block.
    print(img)
```

Knowing that, I was a bit confused [reading the docs](https://fastapi.tiangolo.com/tutorial/request-files/#import-file) and neither seeing a context manager nor a try/finally/else block. something along the lines of:

```python
from fastapi import FastAPI, File, UploadFile

app = FastAPI()


@app.post("/uploadfile/")
async def create_upload_file(file: UploadFile):
    try:
        name = file.filename
    finally:
        file.close()
    return {"filename": name}
```

I then tried finding some part inside the fastapi internals that wraps endpoints in a context manager or a try/finally/else block.  I couldn't. Though that's most certainly my shortcoming of not being very familiar with FastAPI internals :) 

Iis there no danger of memory leakage/file descriptor leakage when not explicitly using UploadFiles `.close()` method inside the endpoint? Would it always be called by Python's memory management?

Note that this question is not meant critically. I've been following FastAPI development for quite some time and I truly appreciate what all the contributors are doing here. My question comes purely from a place of pure paranoia caused by long nights debugging memory leaks throughout the career :) 

### Operating System

macOS

### Operating System Details

macOS 11.6.5 and macOS 12.3.1

### FastAPI Version

0.75.1

### Python Version

3.9.10

**Comment by jgould22** [comment]
FastAPI is built on Starlette. 

You can see the code for that here https://github.com/encode/starlette/blob/master/starlette/datastructures.py#L417

Also you can see some info about it here https://www.starlette.io/requests/

**Comment by originell** [comment]
@jgould22 thanks for taking the time.

I'm afraid this does not answer my question entirely. I might not have been very clear. Let's try again!

I am aware of how UploadFile is implemented in FastAPI and in Starlette. My question is about how memory/disk space is managed when it comes down to using `UploadFile`. Starlette also is not calling `close()` explicitly in its code.

Maybe this example can illustrate my question a bit better: I am uploading a huge file (~1gb). Since all of this is using `SpooledTemporaryFile` internally, the upload should be written to disk right away. 
Usually in Python, once you are done processing, you would call `close()` on that file. This in turn would free up the disk space used.

Now I can't find any place where this `.close()` is called. Is that right? 

If there is no such place, the `close()` is probably called somewhere implicitly? Is it in memory management?

I guess my question is: where is the space (in memory or on disk) freed?

**Comment by jgould22** [comment]
Sorry I did not mean to propose an answer to your question, just adding some context for you.

I am unsure if/where it is freed but you might find more specific answers in the Starlette repo.

**Comment by adriangb** [comment]
As far as I can tell, you are right, FastAPI does not call close, you'd have to do it manually. Not however that UploadFile is buffered in memory for small files, you'd have to send many larger files and see if file descriptors pile up.

**Comment by originell** [comment]
> Sorry I did not mean to propose an answer to your question, just adding some context for you.
> 
> I am unsure if/where it is freed but you might find more specific answers in the Starlette repo.

@jgould22 no need to be sorry! thanks for providing more context!

---

> As far as I can tell, you are right, FastAPI does not call close, you'd have to do it manually. Not however that UploadFile is buffered in memory for small files, you'd have to send many larger files and see if file descriptors pile up.

@adriangb thanks for confirming.

I realize that it takes quite a lot of traffic to exhaust file descriptors etc. Plus, there is also garbage collection that will hopefully also clean up stuff during runtime.

Generally I do feel like it might be better to openly communicate calling `.close()` as a small tipp on how to 'cleanly' take care of handling files without leaving leftovers. I'd be very happy to provide a PR with a suggestion on how I could imagine this looking like in the docs. If that is something that is viewed as a good idea by this community.

Again, thanks for everyone reading and taking the time for providing more context and more insight into everything!

**Comment by adriangb** [comment]
I don't think it's that simple for FastAPI: the files are opened before the request handler is called, and unrelated error, like deseralizing another form field, could cause an exception so that you never even have the opportunity to close the file.

Opened here: https://github.com/tiangolo/fastapi/blob/146f57b8f70c5757dc20edc716dba1b96936a8d6/fastapi/routing.py#L192

But it's totally possible that your request handler never gets called so you'd have no chance to close the file manually

**Comment by originell** [comment, marked as answer]
Thanks for the deep dive. So that means we simply rely on memory management/garbage collection to take care of stuff :) good to know! thanks for the. clarification
